File: typescript/src/path/allowlist.ts
// Module-level singleton for the path-confinement allow-list.
//
// cli.ts builds the allow-list once at startup via buildAllowList(argv) and
// stores the resolved roots here. The store_artifact path branch reads them via
// getAllowRoots() at handler time and hands them to checkPath(). Keeping the
// roots in a tiny module (mirroring http/instance.ts) avoids threading them
// through the Server instance or every tool signature.

let _roots: string[] | undefined;

export function setAllowRoots(roots: string[]): void {
  _roots = roots;
}

export function getAllowRoots(): string[] {
  if (!_roots) {
    throw new Error(
      "Path allow-list not initialised. Call setAllowRoots() at startup before invoking path-based tools."
    );
  }
  return _roots;
}

export function resetAllowRoots(): void {
  _roots = undefined;
}

File: python/src/artifacta_mcp/allowlist.py
"""Module-level singleton for the path-confinement allow-list.

Mirrors `mcp/typescript/src/path/allowlist.ts`. cli.py calls
`build_allow_list(argv)` once at startup, then stores the resolved roots
via `set_allow_roots`. The store_artifact path branch reads them with
`get_allow_roots()` at handler time.
"""
from __future__ import annotations

_roots: list[str] | None = None


def set_allow_roots(roots: list[str]) -> None:
    global _roots
    _roots = list(roots)


def get_allow_roots() -> list[str]:
    if _roots is None:
        raise RuntimeError(
            "Path allow-list not initialised. Call set_allow_roots() at startup "
            "before invoking path-based tools."
        )
    return _roots


def reset_allow_roots() -> None:
    global _roots
    _roots = None
